pkg/syncer/storagepool/intended_state.go

```diff
@@ -211,23 +217,44 @@ func newIntendedVsanSNAState(ctx context.Context, scWatchCntlr *StorageClassWatc
 			compatSC = append(compatSC, scName)
 		}
 	}
+	// In vSAN-SNA overhead is upperbound by roughly 4% of volume requested capacity. Thus maximum allocatable space in vSAN-SNA is 96% of free-capacity.
+	vsanSnaDsType := fmt.Sprintf("%s-sna", vsan.dsType)
+	freeSpace := resource.NewQuantity(vsanHost.Capacity - vsanHost.CapacityUsed, resource.DecimalSI)
+	allocatableSpace := getAllocatableSpace(freeSpace, vsanSnaDsType)
 
 	return &intendedState{
-		dsMoid:        fmt.Sprintf("%s-%s", vsan.dsMoid, node),
-		dsType:        fmt.Sprintf("%s-sna", vsan.dsType),
-		spName:        fmt.Sprintf("%s-%s", vsan.spName, node),
-		capacity:      resource.NewQuantity(vsanHost.Capacity, resource.DecimalSI),
-		freeSpace:     resource.NewQuantity(vsanHost.Capacity-vsanHost.CapacityUsed, resource.DecimalSI),
-		url:           vsan.url,
-		accessible:    true,  // If this node is in accessible node list of the vsan-ds, then sp is accessible
-		datastoreInMM: false, // If this node is inMM - the storagepool will not exist at all
-		allHostsInMM:  false, // If this node is inMM - the storagepool will not exist at all
-		nodes:         nodes,
-		compatSC:      compatSC,
-		isRemoteVsan:  false,
+		dsMoid:           fmt.Sprintf("%s-%s", vsan.dsMoid, node),
+		dsType:           vsanSnaDsType,
+		spName:           fmt.Sprintf("%s-%s", vsan.spName, node),
+		capacity:         resource.NewQuantity(vsanHost.Capacity, resource.DecimalSI),
+		freeSpace:        freeSpace,
+		allocatableSpace: allocatableSpace,
+		url:              vsan.url,
+		accessible:       true,  // If this node is in accessible node list of the vsan-ds, then sp is accessible
+		datastoreInMM:    false, // If this node is inMM - the storagepool will not exist at all
+		allHostsInMM:     false, // If this node is inMM - the storagepool will not exist at all
+		nodes:            nodes,
+		compatSC:         compatSC,
+		isRemoteVsan:     false,
 	}, nil
 }
 
+func getAllocatableSpace(freeSpace *resource.Quantity, dsType string) *resource.Quantity {
+	if dsType == spTypePrefix+"vsanD" {
+		// In vSAN Direct overhead to provision volume is upperbound by 4MiB space. Thus maximum allocatable space in vSAN Direct is 4MiB less than free-capacity.
+		allocatableSpaceInBytes := freeSpace.Value() - 4*1024*1024
+		if allocatableSpaceInBytes > 0 {
+			return resource.NewQuantity(allocatableSpaceInBytes, resource.DecimalSI)
+		}
+		return resource.NewQuantity(0, resource.DecimalSI)
+	}
+	// In vSAN-SNA overhead is upperbound by roughly 4% of volume requested capacity. Thus maximum allocatable space in vSAN-SNA is 96% of free-capacity.
```

Comment:
What is the minimum disk size for vSAN SNA?

Since we are using a % of disk size, should we calculate the overhead = max (4% of disk size, a fixed value like 4mb)?

Reply:
yes there could be a fixed overhead in vSAN SNA along with a proportionate overhead with volume size. We really don't know what this number is. For sure its not a huge number. for eg. if the fixed overhead was 4 MB then this fixed overhead would significant only if the size of vsan SAN disks are about 200 MB. In the customer environment vsan SNA total capacities are in TBs. We will hit this scenario only when these disk are 99.98% occupied. vSAN recommends to not overfill the vSAN cluster above 95% to not affect maintenance tasks. We can safely say that for all practical purposes using 4% of disk size as max overhead suffices.